man/Conformal3Proj.1
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Pod::Man v6.0.2 (Pod::Simple 3.45)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Required to disable full justification in groff 1.23.0.
.if n .ds AD l
.\" ========================================================================
.\"
.IX Title "CONFORMAL3PROJ 1"
.TH CONFORMAL3PROJ 1 2025-11-05 "GeographicLib 2.7" "GeographicLib Utilities"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH NAME
Conformal3Proj \-\- conformal projections for a triaxial ellipsoid
.SH SYNOPSIS
.IX Header "SYNOPSIS"
\&\fBConformal3Proj\fR [ \fB\-r\fR ]
[ \fB\-t\fR \fIa\fR \fIb\fR \fIc\fR | \fB\-e\fR \fIb\fR \fIe2\fR \fIk2\fR \fIkp2\fR ]
[ \fB\-tx\fR \fIax\fR \fIbx\fR \fIcx\fR | \fB\-ex\fR \fIbx\fR \fIe2x\fR \fIk2x\fR \fIkp2x\fR ]
[ \fB\-d\fR | \fB\-:\fR ] [ \fB\-w\fR ] [ \fB\-p\fR \fIprec\fR ]
[ \fB\-\-comment\-delimiter\fR \fIcommentdelim\fR ]
[ \fB\-\-version\fR | \fB\-h\fR | \fB\-\-help\fR ]
[ \fB\-\-input\-file\fR \fIinfile\fR | \fB\-\-input\-string\fR \fIinstring\fR ]
[ \fB\-\-line\-separator\fR \fIlinesep\fR ]
[ \fB\-\-output\-file\fR \fIoutfile\fR ]
.SH DESCRIPTION
.IX Header "DESCRIPTION"
Project points using the Jacobi conformal projection.  By default, the
coordinates of points on a triaxial ellipsoid are read from standard
input as ellipsoidal latitude and longitude and the projected points are
written to standard output as easting and northing.  The scale of the
projection is appended to the output.  Specifying \fB\-r\fR causes the
reverse projection to be performed.
.PP
If either \fB\-tx\fR or \fB\-ex\fR is given, then this describes a target
ellipsoid onto which the points are conformally mapped via the Jacobi
projections to intermediate spheres.  In this case, the meridian
convergence and the scale are appended to the output.
.SH OPTIONS
.IX Header "OPTIONS"
.IP \fB\-r\fR 4
.IX Item "-r"
perform the reverse projection.
.IP "\fB\-t\fR \fIa\fR \fIb\fR \fIc\fR" 4
.IX Item "-t a b c"
specify the ellipsoid via its major semiaxis \fIa\fR, median semiaxis \fIb\fR,
and minor semixis \fIc\fR.  By default, we have \fIa\fR = 6378172 m, \fIb\fR =
6378102 m, \fIc\fR = 6356752 m, an approximate triaxial model of the earth.
(With this model \fIomg\fR = 0deg, corresponds to \fIlon\fR = \-14.93deg.
.IP "\fB\-e\fR \fIb\fR \fIe2\fR \fIk2\fR \fIkp2\fR" 4
.IX Item "-e b e2 k2 kp2"
specify the ellipsoid via the median semiaxis, \fIb\fR and the shape
parameters \fIe2\fR = (\fIa\fR^2 \- \fIc\fR^2)/\fIb\fR^2, \fIk2\fR = (\fIb\fR^2 \-
\&\fIc\fR^2)/(\fIa\fR^2 \- \fIc\fR^2), and \fIkp2\fR = (\fIa\fR^2 \- \fIb\fR^2)/(\fIa\fR^2 \-
\&\fIc\fR^2).  Simple fractions are allowed for \fIe2\fR, \fIk2\fR, and \fIkp2\fR.
Internally, the supplied values of \fIk2\fR and \fIkp2\fR are normalized so
that \fIk2\fR + \fIkp2\fR = 1.
.IP "\fB\-tx\fR \fIax\fR \fIbx\fR \fIcx\fR" 4
.IX Item "-tx ax bx cx"
specify the target ellipsoid via its semiaxes.
.IP "\fB\-ex\fR \fIbx\fR \fIe2x\fR \fIk2x\fR \fIkp2x\fR" 4
.IX Item "-ex bx e2x k2x kp2x"
specify the target ellipsoid via its median semiaxes and shape
parameters.
.IP \fB\-d\fR 4
.IX Item "-d"
output angles as degrees, minutes, seconds instead of decimal degrees.
.IP \fB\-:\fR 4
.IX Item "-:"
like \fB\-d\fR, except use : as a separator instead of the d, \*(Aq, and "
delimiters.
.IP \fB\-w\fR 4
.IX Item "-w"
toggle the longitude first flag (it starts off); if the flag is on, then
on input and output, longitude precedes latitude (except that, on input,
this can be overridden by a hemisphere designator, \fIN\fR, \fIS\fR, \fIE\fR,
\&\fIW\fR).
.IP "\fB\-p\fR \fIprec\fR" 4
.IX Item "-p prec"
set the output precision to \fIprec\fR (default 3).  For distances, \fIprec\fR
is the number of digits after the decimal point for ellipsoids which are
approximately the same size as the Earth; for other ellipsoids the
precision is adjusted to retain the same relative precision.  For
latitudes and longitudes (in degrees), the number of digits after the
decimal point is \fIprec\fR + 5.  For the scale, the precision is \fIprec\fR +
7.
.IP "\fB\-\-comment\-delimiter\fR \fIcommentdelim\fR" 4
.IX Item "--comment-delimiter commentdelim"
set the comment delimiter to \fIcommentdelim\fR (e.g., "#" or "//").  If
set, the input lines will be scanned for this delimiter and, if found,
the delimiter and the rest of the line will be removed prior to
processing and subsequently appended to the output line (separated by a
space).
.IP \fB\-\-version\fR 4
.IX Item "--version"
print version and exit.
.IP \fB\-h\fR 4
.IX Item "-h"
print usage and exit.
.IP \fB\-\-help\fR 4
.IX Item "--help"
print full documentation and exit.
.IP "\fB\-\-input\-file\fR \fIinfile\fR" 4
.IX Item "--input-file infile"
read input from the file \fIinfile\fR instead of from standard input; a file
name of "\-" stands for standard input.
.IP "\fB\-\-input\-string\fR \fIinstring\fR" 4
.IX Item "--input-string instring"
read input from the string \fIinstring\fR instead of from standard input.
All occurrences of the line separator character (default is a semicolon)
in \fIinstring\fR are converted to newlines before the reading begins.
.IP "\fB\-\-line\-separator\fR \fIlinesep\fR" 4
.IX Item "--line-separator linesep"
set the line separator character to \fIlinesep\fR.  By default this is a
semicolon.
.IP "\fB\-\-output\-file\fR \fIoutfile\fR" 4
.IX Item "--output-file outfile"
write output to the file \fIoutfile\fR instead of to standard output; a
file name of "\-" stands for standard output.
.SH EXAMPLES
.IX Header "EXAMPLES"
.Vb 4
\&   echo 33.3 44.4 | Conformal3Proj
\&   \-5077732.396 3922571.859 1.1970343759
\&   echo \-5077732.396 3922571.859 | Conformal3Proj \-r
\&   33.30000000 44.40000000 1.1970343759
.Ve
.SH ERRORS
.IX Header "ERRORS"
An illegal line of input will print an error message to standard output
beginning with \f(CW\*(C`ERROR:\*(C'\fR and causes \fBConformal3Proj\fR to return an exit
code of 1.  However, an error does not cause \fBConformal3Proj\fR to
terminate; following lines will be converted.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
Ellipsoidal coordinates are documented in Section 2 of C. F. F. Karney,
\&\fIJacobi\*(Aqs solution for geodesics on a triaxial ellipsoid\fR,
Technical Report, SRI International, Nov. 2025,
<https://arxiv.org/abs/2511.nnnnn>.
.PP
The Wikipedia page, Geodesics on an ellipsoid,
<https://en.wikipedia.org/wiki/Geodesics_on_an_ellipsoid>.
.SH AUTHOR
.IX Header "AUTHOR"
\&\fBConformal3Proj\fR was written by Charles Karney.
.SH HISTORY
.IX Header "HISTORY"
\&\fBConformal3Proj\fR was added to GeographicLib,
<https://geographiclib.sourceforge.io>, in version 2.6.


NEWS
A reverse chronological list of changes to GeographicLib

For more information, see

    https://geographiclib.sourceforge.io/C++/doc/

The current version of the library is 2.7.

Changes between 2.7 (released 2025-11-06) and 2.6 versions:

   * Fix two bugs in classes for triaxial ellipsoids:
     + Triaxial::Geodesic3::Inverse sometimes threw an exception when
       both endpoints were on the major principal ellipse for very
       eccentric ellipsoids a > 2 c.
     + Triaxial::Cartesian3::carttocart2 sometimes returned the wrong
       result in the interior of the ellipsoid with Z = 0.

   * Cite the paper covering some of the triaxial classes Jacobi's
     solution for geodesics on a triaxial ellipsoid.

   * Minor documentation fixes.

Changes between 2.6 (released 2025-09-30) and 2.5.2 versions:

   * Support for triaxial ellipsoids. This is all placed in the
     GeographicLib::Triaxial namespace with the header files in the
     GeographicLib/Triaxial directory. This provides
     + Triaxial::Ellipsoid3 supports basic conversions of ellipsoidal
       coordinates.
     + Triaxial::GeodesicLine3 and Triaxial::Geodesic3 solve the direct
       and inverse geodesic problems on a triaxial ellipsoid. This
       implements Jacobi's solution for the direct problem. The inverse
       problem extends the method for solving the inverse problem on a
       biaxial ellipsoid.
     + Triaxial::Cartesian3 performs conversions between different
       systems of triaxial coordinates.
     + Triaxial::Conformal3 implements Jacobi's conformal projection for
       a triaxial ellipsoid. This allow a triaxial ellipsoid to be
       conformally mapped to a sphere or to any other ellipsoid.
     + Geod3Solve is a command-line utility offering access to
       Triaxial::Geodesic3.
     + Cart3Convert is a command-line utility offering access to
       Triaxial::Cartesian3.
     + Conformal3Proj is a command-line utility offering access to
       Triaxial::Conformal3.
     + Constants::Triaxial_Earth_a, Constants::Triaxial_Earth_b,
       Constants::Triaxial_Earth_c, and Constants::Triaxial_Earth_lon0
       specify a simple triaxial model for the Earth.
     + experimental::TriaxialGeodesicODE is provided to illustrate the
       solution of the direct geodesic problem by integrating the
       ordinary differential equations for the geodesic. Because this
       depends on Boost, this is not part of the GeographicLib library.
     + experimental/Geod3ODE.cpp a utility for doing direct geodesic
       calculations using experimental::TriaxialGeodesicODE.

   * To support the triaxial routines, the following classes have been
     added:
     + AngleT provides an accurate representation of angles of any
       magnitude. This might replace AuxAngle at some point.
     + Trigfun supports representing periodic functions as Fourier
       series with automatic selection of the number of terms to
       use. This offers a tiny subset of the functionality of the trig
       series in Chebfun. This might replace DST at some point.

   * C++17 is now required. Minimum version of Visual Studio supported
     is Visual Studio 15 2017.

   * Add Math::clamp.

   * Because experimental::TriaxialGeodesicODE depends on Boost, the
     CMake option USE_BOOST_FOR_EXAMPLES has been renamed
     USE_BOOST. This will cause find_package(Boost) to be called, but
     does not affect how the library or the tools are built.

Changes between 2.5.2 (released 2025-08-22) and 2.5.1 versions:

   * Fix potential buffer overflow in DMS.  Upgrading from earlier
     versions is strongly recommended.  If upgrading is not possible,
     avoid processing untrusted or malformed inputs and run programs
     linked to the library with restricted privileges.  Thanks to
     Rosario Matteo Grammatico for finding this vulnerability.

Changes between 2.5.1 (released 2025-08-20) and 2.5 versions:

   * Fix bug in geodesic inverse problem with very prolate
     ellipsoids. For f < -2, a meridional path was incorrectly
     returned in some cases for two points on the same meridian. This
     affected both the series and exact solutions. (Or course, the
     series solution is inaccurate for ellipsoids this eccentric.)

   * Minor tweak to Math::atan2d to improve accuracy.

   * Allow GEOGRAPHICLIB_PRECISION > 5 to allow use of mpreal with a
     fixed precision.

   * Document that cmake option -D BINDIR=OFF removes tools from the
     default all target for make.

   * WARNING: C++17 compliant compiler will be required for the next
     release. Support for builds with Visual Studio 14 (2015) will
     then stop.

Changes between 2.5 (released 2024-12-28) and 2.4 versions:

   * New magnetic models provided

     + World Magnetic Model 2025, wmm2025; this covers the period
       2025-2030.  This is now the model returned by
       MagneticModel::DefaultMagneticName and is default magnetic
       model for MagneticField (replacing wmm2020 which is only valid
       thru the end of 2024).
     + World Magnetic Model High Resolution 2025, wmmhr2025; this is a
       high-resolution version of wmm2025 including terms up to degree
       133 (as opposed to 12 for wmm2025).
     + International Geomagnetic Reference Field (14th generation),
       igrf14, which approximates the main magnetic field of the earth
       for the period 1900-2030.

   * Update CMake's cmake_minimum_required to 3.17.0 (to facilitate
     integration with Qt projects).

   * For GEOGRAPHICLIB_PRECISION > 2, readarray treats doubles as
     approximations to exact decimal constants.

Changes between 2.4 (released 2024-07-14) and 2.3 versions:

   * Math::sind, Math::sincosd, etc., now return accurate results for
     arguments = 30d, 45d, 60d, ...

   * Add Math::hypot3.

   * Add EllipticFunction::am.

   * Make constructor for AuxAngle explicit.

   * Include notes for plate carree areas.

   * C++14 compliant compiler required (this requirement tracks the

tools/Conformal3Proj.cpp
/**
 * \file Conformal3Proj.cpp
 * \brief Command line utility for computing geodesics on a triaxial ellipsoid
 *
 * Copyright (c) Charles Karney (2024-2025) <karney@alum.mit.edu> and licensed
 * under the MIT/X11 License.  For more information, see
 * https://geographiclib.sourceforge.io/
 *
 * See the <a href="Conformal3Proj.1.html">man page</a> for usage information.
 **********************************************************************/

// Usual flags
//   -r reverse
//   -e (supplement with -t)
//   -w longfirst
//   -p prec
//   -h help
//   -d dms
//   -: colon
//   --help
//   --version
//   --comment-delimiter
//   + other I/O related flags

// New flags
//   -ex -tx

#include <iostream>
#include <iomanip>
#include <string>
#include <sstream>
#include <fstream>
#include <GeographicLib/DMS.hpp>
#include <GeographicLib/Utility.hpp>
#include <GeographicLib/Angle.hpp>
#include <GeographicLib/Triaxial/Conformal3.hpp>

#include "Conformal3Proj.usage"

int main(int argc, const char* const argv[]) {
  try {
    using namespace GeographicLib;
    using namespace Triaxial;
    using real = Math::real;
    using ang = Angle;
    Utility::set_digits();
    bool reverse = false, dms = false,
      longfirst = false, altell = false;
    real
      a = Constants::Triaxial_Earth_a(),
      b = Constants::Triaxial_Earth_b(),
      c = Constants::Triaxial_Earth_c(),
      e2 = -1, k2 = 0, kp2 = 0,
      ax = -1, bx = 0, cx = 0, e2x = -1,  k2x = 0, kp2x = 0;
    int prec = 3;
    std::string istring, ifile, ofile, cdelim;
    char lsep = ';', dmssep = char(0);

    for (int m = 1; m < argc; ++m) {
      std::string arg(argv[m]);
      if (arg == "-r")
        reverse = true;
      else if (arg == "-t") {
        if (m + 3 >= argc) return usage(1, true);
        try {
          a = Utility::val<real>(std::string(argv[m + 1]));
          b = Utility::val<real>(std::string(argv[m + 2]));
          c = Utility::val<real>(std::string(argv[m + 3]));
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding arguments of -t: " << e.what() << "\n";
          return 1;
        }
        e2 = -1;
        m += 3;
      } else if (arg == "-e") {
        // Cayley ellipsoid sqrt([2,1,1/2]) is
        // -e 1 3/2 1/3 2/3
        if (m + 4 >= argc) return usage(1, true);
        try {
          b = Utility::val<real>(std::string(argv[m + 1]));
          e2 = Utility::fract<real>(std::string(argv[m + 2]));
          k2 = Utility::fract<real>(std::string(argv[m + 3]));
          kp2 = Utility::fract<real>(std::string(argv[m + 4]));
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding arguments of -e: " << e.what() << "\n";
          return 1;
        }
        a = -1;
        m += 4;
      } else if (arg == "-tx") {
        if (m + 3 >= argc) return usage(1, true);
        try {
          ax = Utility::val<real>(std::string(argv[m + 1]));
          bx = Utility::val<real>(std::string(argv[m + 2]));
          cx = Utility::val<real>(std::string(argv[m + 3]));
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding arguments of -t: " << e.what() << "\n";
          return 1;
        }
        e2x = -1;
        m += 3;
      } else if (arg == "-ex") {
        // Cayley ellipsoid sqrt([2,1,1/2]) is
        // -e 1 3/2 1/3 2/3
        if (m + 4 >= argc) return usage(1, true);
        try {
          bx = Utility::val<real>(std::string(argv[m + 1]));
          e2x = Utility::fract<real>(std::string(argv[m + 2]));
          k2x = Utility::fract<real>(std::string(argv[m + 3]));
          kp2x = Utility::fract<real>(std::string(argv[m + 4]));
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding arguments of -e: " << e.what() << "\n";
          return 1;
        }
        ax = -1;
        m += 4;
      } else if (arg == "-d") {
        dms = true;
        dmssep = '\0';
      } else if (arg == "-:") {
        dms = true;
        dmssep = ':';
      } else if (arg == "-w")
        longfirst = !longfirst;
      else if (arg == "-p") {
        if (++m == argc) return usage(1, true);
        try {
          prec = Utility::val<int>(std::string(argv[m]));
        }
        catch (const std::exception&) {
          std::cerr << "Precision " << argv[m] << " is not a number\n";
          return 1;
        }
      } else if (arg == "--input-string") {
        if (++m == argc) return usage(1, true);
        istring = argv[m];
      } else if (arg == "--input-file") {
        if (++m == argc) return usage(1, true);
        ifile = argv[m];
      } else if (arg == "--output-file") {
        if (++m == argc) return usage(1, true);
        ofile = argv[m];
      } else if (arg == "--line-separator") {
        if (++m == argc) return usage(1, true);
        if (std::string(argv[m]).size() != 1) {
          std::cerr << "Line separator must be a single character\n";
          return 1;
        }
        lsep = argv[m][0];
      } else if (arg == "--comment-delimiter") {
        if (++m == argc) return usage(1, true);
        cdelim = argv[m];
      } else if (arg == "--version") {
        std::cout << argv[0] << ": GeographicLib version "
                  << GEOGRAPHICLIB_VERSION_STRING << "\n";
        return 0;
      } else
        return usage(!(arg == "-h" || arg == "--help"), arg != "--help");
    }

    Conformal3 tc(e2 >= 0 ? Ellipsoid3(b, e2, k2, kp2) : Ellipsoid3(a, b, c));
    altell = e2x >= 0 || ax > 0;
    Conformal3 tcx(e2x >= 0 ? Ellipsoid3(bx, e2x, k2x, kp2x) :
                   ax > 0 ? Ellipsoid3(ax, bx, cx) : Ellipsoid3());

    if (!ifile.empty() && !istring.empty()) {
      std::cerr << "Cannot specify --input-string and --input-file together\n";
      return 1;
    }
    if (ifile == "-") ifile.clear();
    std::ifstream infile;
    std::istringstream instring;
    if (!ifile.empty()) {
      infile.open(ifile.c_str());
      if (!infile.is_open()) {
        std::cerr << "Cannot open " << ifile << " for reading\n";
        return 1;
      }
    } else if (!istring.empty()) {
      std::string::size_type m = 0;
      while (true) {
        m = istring.find(lsep, m);
        if (m == std::string::npos)
          break;
        istring[m] = '\n';
      }
      instring.str(istring);
    }
    std::istream* input = !ifile.empty() ? &infile :
      (!istring.empty() ? &instring : &std::cin);

    std::ofstream outfile;
    if (ofile == "-") ofile.clear();
    if (!ofile.empty()) {
      outfile.open(ofile.c_str());
      if (!outfile.is_open()) {
        std::cerr << "Cannot open " << ofile << " for writing\n";
        return 1;
      }
    }
    std::ostream* output = !ofile.empty() ? &outfile : &std::cout;

    // Max precision = 10: 0.1 nm in distance, 10^-15 deg (= 0.11 nm),
    // 10^-11 sec (= 0.3 nm).
    prec = std::min(10 + Math::extra_digits(), std::max(0, prec));
    using std::round, std::log10, std::ceil, std::signbit;
    int disprec = std::max(0, prec + int(round(log10(6400000/b)))),
      angprec = prec + 5, scalprec = prec + 7;
    if (altell) {
      scalprec += int(round(log10(b/bx)));
      scalprec = std::max(0, scalprec);
  }
    std::string s, eol, stra, strb, strc;
    std::istringstream str;
    int retval = 0;
    while (std::getline(*input, s)) {
      try {
        eol = "\n";
        if (!cdelim.empty()) {
          std::string::size_type m = s.find(cdelim);
          if (m != std::string::npos) {
            eol = " " + s.substr(m) + "\n";
            s = s.substr(0, m);
          }
        }
        // READ
        str.clear(); str.str(s);
        if (!(str >> stra >> strb))
          throw GeographicErr("Incomplete input: " + s);
        if (str >> strc)
          throw GeographicErr("Extraneous input: " + strc);
        ang gam{}; real k = 1;
        if (reverse) {
          ang bet, omg;
          if (altell) {
            ang betx, omgx;
            ang::DecodeLatLon(stra, strb, betx, omgx, longfirst);
            tc.ReverseOther(tcx, betx, omgx, bet, omg, gam, k);
          } else {
            real x = Utility::val<real>(stra), y = Utility::val<real>(strb);
            tc.Reverse(x, y, bet, omg, k);
          }
          *output << ang::LatLonString(bet, omg,
                                       angprec, dms, dmssep, longfirst);
        } else {
          ang bet, omg;
          ang::DecodeLatLon(stra, strb, bet, omg, longfirst);
          if (altell) {
            ang betx, omgx;
            tc.ForwardOther(tcx, bet, omg, betx, omgx, gam, k);
            *output << ang::LatLonString(betx, omgx,
                                         angprec, dms, dmssep, longfirst);
          } else {
            real x, y;
            tc.Forward(bet, omg, x, y, k);
            *output << Utility::str(x, disprec) << " "
                    << Utility::str(y, disprec);
          }
        }
        if (altell)
          *output << " " << ang::AzimuthString(gam, angprec, dms, dmssep);
        *output << " " << Utility::str(k, scalprec) << eol;
      }
      catch (const std::exception& e) {
        // Write error message cout so output lines match input lines
        *output << "ERROR: " << e.what() << "\n";
        retval = 1;
      }
    }
    return retval;
  }
  catch (const std::exception& e) {
    std::cerr << "Caught exception: " << e.what() << "\n";
    return 1;
  }
  catch (...) {
    std::cerr << "Caught unknown exception\n";
    return 1;
  }
}
